compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -821,17 +821,31 @@ object PatternMatcher {
         }
       }
 
-    /** Emit cases of a switch */
-    private def emitSwitchCases(cases: List[(List[Tree], Plan)]): List[CaseDef] = (cases: @unchecked) match {
-      case (alts, ons) :: cases1 =>
+    /** Emit a switch-match */
+    private def emitSwitchMatch(scrutinee: Tree, cases: List[(List[Tree], Plan)]): Match = {
+      /* Make sure to adapt the scrutinee to Int, as well as all the alternatives
+       * of all cases, so that only Matches on pritimive Ints survive this phase.
+       */
+
+      val intScrutinee =
+        if (scrutinee.tpe.widen.isRef(defn.IntClass)) scrutinee
+        else scrutinee.select(nme.toInt)
+
+      def intLiteral(lit: Tree): Tree =
+        val Literal(constant) = lit
+        if (constant.tag == Constants.IntTag) lit
+        else Literal(Constant(constant.intValue)).withSpan(lit.span)
```

Comment:
```suggestion
        else cpy.Literal(lit)(Constant(constant.intValue))
```
to also set the source file of the tree corectly

Reply:
Thank you, that's much better!

Updated.